### PR TITLE
feat(git): add git Bitbucket provider

### LINKED ISSUE #86: Support BitBucket as a git provider
Parent: https://github.com/fabric8-launcher/launcher-planning/issues/93

### PR DESCRIPTION
**Pull request by ia3andy**
Some issues have been created to fix some related problems.

Fixes #86

### CODE REVIEW

**Comment by gastaldi** [comment]
Add to whitelist

File: README.md [hunk 1]
```diff
@@ -125,7 +125,7 @@ Prerequisites to Run Integration Tests
     * Install minishift and prerequisite projects by following these instructions
         * https://github.com/minishift/minishift#installing-minishift
 	
-    * Check everything works okay by loggin in to the OpenShift console
+    * Check everything works okay by login in to the OpenShift console
```
Comment (quintesse) [inline README.md]: Both the original and your fix are wrong ;)

Should be "logging"

File: README.md [hunk 2]
```diff
@@ -233,6 +233,8 @@ export LAUNCHER_MISSIONCONTROL_GITHUB_USERNAME=`git config github.user`
 export LAUNCHER_MISSIONCONTROL_GITHUB_TOKEN=`git config github.token`
 export LAUNCHER_MISSIONCONTROL_GITLAB_USERNAME=`git config gitlab.user`
 export LAUNCHER_MISSIONCONTROL_GITLAB_PRIVATE_TOKEN=`git config gitlab.token`
+export LAUNCHER_MISSIONCONTROL_BITBUCKET_USERNAME=BitbucketFakeUserName
+export LAUNCHER_MISSIONCONTROL_BITBUCKET_APPLICATION_PASSWORD=BitbucketFakeApplicationPassword1234$
```
Comment (quintesse) [inline README.md]: Why this? Why not the same as the others?

**Comment by gastaldi** [comment]
By the way it's a good practice to not merge our own PRs, but let someone from the team to do it after reviewing 😉

File: README.md (hunk 2 above)
Comment (quintesse) [inline README.md]: We're also missing a section on setting up a BitBucket account here: https://github.com/ia3andy/launcher-backend/blob/fc1bffdb618f32fe12220cc1bc523b041c50da07/README.md#prerequisites-to-run-integration-tests

File: base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java [hunk 3]
```diff
@@ -1,10 +1,14 @@
 package io.fabric8.launcher.base.identity;
 
+import java.util.Base64;
+
 /**
  * @author <a href="mailto:ggastald@redhat.com">George Gastaldi</a>
  */
 public class UserPasswordIdentity implements Identity {
 
+    private static final String BASIC_PREFIX = "Basic ";
```
Comment (gastaldi) [inline base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java]: Where is this field used?

File: services/git-service-impl/src/main/java/io/fabric8/launcher/service/bitbucket/impl/BitbucketServiceFactoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+import io.fabric8.launcher.base.identity.TokenIdentity;
+import io.fabric8.launcher.base.identity.UserPasswordIdentity;
+import io.fabric8.launcher.service.bitbucket.api.BitbucketService;
+import io.fabric8.launcher.service.bitbucket.api.BitbucketServiceFactory;
+
+public class BitbucketServiceFactoryImpl implements BitbucketServiceFactory {
+    @Override
+    public BitbucketService create(final Identity identity) {
+        if (!(identity instanceof UserPasswordIdentity)) {
+            throw new IllegalArgumentException("BitbucketService supports only TokenIdentity. Not supported:" + identity);
```
Comment (gastaldi) [inline services/git-service-impl/src/main/java/io/fabric8/launcher/service/bitbucket/impl/BitbucketServiceFactoryImpl.java]: This error message does not look right. Or maybe the if statement is wrong

**Comment by ia3andy** [comment]
I thought "Add to Whitelist" meant that it was ready to merge.. I'll fix those on monday.. Sorry about that.

**Comment by gastaldi** [comment]
No, that means: CI server, trust this collaborator 😃

File: services/git-service-impl/src/main/java/io/fabric8/launcher/service/bitbucket/impl/BitbucketServiceFactoryImpl.java (hunk 4 above)
Comment (ia3andy) [reply services/git-service-impl/src/main/java/io/fabric8/launcher/service/bitbucket/impl/BitbucketServiceFactoryImpl.java]: Good catch :) it's only UserPasswordIdentity

File: base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java (hunk 3 above)
Comment (ia3andy) [reply base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java]: Indeed, it has been moved..

File: base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java [hunk 5]
```diff
@@ -1,10 +1,14 @@
 package io.fabric8.launcher.base.identity;
 
+import java.util.Base64;
```
[the remark this answers is not in the document]
Comment (ia3andy) [reply base/src/main/java/io/fabric8/launcher/base/identity/UserPasswordIdentity.java]: Indeed, it has been moved

File: README.md (hunk 2 above)
Comment (ia3andy) [reply README.md]: It's not a token, it's a Application Password as they call it. It's used with basic auth, not as a Bearer token.

**Pull request merged**